include/loader/serialize.h

```diff
@@ -23,56 +25,180 @@ namespace Loader {
+  Expect<void> serializeType(const AST::MemoryType &Type, std::vector<uint8_t> &OutVec);
+  Expect<void> serializeType(const AST::GlobalType &Type, std::vector<uint8_t> &OutVec);
+  Expect<void> serializeExpression(const AST::Expression &Expr,
                            std::vector<uint8_t> &OutVec);
-  void serializeInstruction(const AST::Instruction &Instr,
+  Expect<void> serializeInstruction(const AST::Instruction &Instr,
                             std::vector<uint8_t> &OutVec);
   /// @}
 
   /// \name Helper functions
   /// @{
-  void serializeU32(uint32_t Num, std::vector<uint8_t> &OutVec,
-                    std::vector<uint8_t>::iterator It) {
-    uint8_t Buf[5];
+  inline auto logSerializeError(ErrCode Code, ASTNodeAttr Node) const noexcept {
+    spdlog::error(Code);
+    spdlog::error(ErrInfo::InfoAST(Node));
+    return Unexpect(Code);
+  }
+  inline auto logNeedProposal(ErrCode Code, Proposal Prop, ASTNodeAttr Node) const noexcept {
+    spdlog::error(Code);
+    spdlog::error(ErrInfo::InfoProposal(Prop));
+    spdlog::error(ErrInfo::InfoAST(Node));
+    return Unexpect(Code);
+  }
+
+  // Helper function of checking the valid value types.
+  Expect<ValType> checkValTypeProposals(ValType VType, ASTNodeAttr Node) const noexcept {
+    if (VType == ValType::V128 && !Conf.hasProposal(Proposal::SIMD)) {
+      return logNeedProposal(ErrCode::Value::MalformedValType, Proposal::SIMD, Node);
+    }
+    if ((VType == ValType::FuncRef &&
+         !Conf.hasProposal(Proposal::ReferenceTypes) &&
+         !Conf.hasProposal(Proposal::BulkMemoryOperations)) ||
+        (VType == ValType::ExternRef &&
+         !Conf.hasProposal(Proposal::ReferenceTypes))) {
+      return logNeedProposal(ErrCode::Value::MalformedElemType,
+                             Proposal::ReferenceTypes, Node);
+    }
+    switch (VType) {
+    case ValType::I32:
+    case ValType::I64:
+    case ValType::F32:
+    case ValType::F64:
+    case ValType::V128:
+    case ValType::ExternRef:
+    case ValType::FuncRef:
+      return VType;
+    default:
+      return logSerializeError(ErrCode::Value::MalformedValType, Node);
+    }
+  }
+
+  // Helper function of checking the valid reference types.
+  Expect<RefType> checkRefTypeProposals(RefType RType, ASTNodeAttr Node) const noexcept {
+    switch (RType) {
+    case RefType::ExternRef:
+      if (!Conf.hasProposal(Proposal::ReferenceTypes)) {
+        return logNeedProposal(ErrCode::Value::MalformedElemType,
+                               Proposal::ReferenceTypes, Node);
+      }
+      [[fallthrough]];
+    case RefType::FuncRef:
+      return RType;
+    default:
+      if (Conf.hasProposal(Proposal::ReferenceTypes)) {
+        return logSerializeError(ErrCode::Value::MalformedRefType, Node);
+      } else {
+        return logSerializeError(ErrCode::Value::MalformedElemType, Node);
+      }
+    }
+  }
+
+  Expect<void> checkInstrProposals(OpCode Code) const noexcept {
```

Comment:
The proposal checking functions are copied from loader.
Maybe you can refine them and make them as member functions as `Configure`?